File: app/src/main/java/ch/bailu/aat/gpx/tools/TimeStampFixer.java
package ch.bailu.aat.gpx.tools;

import ch.bailu.aat.gpx.AltitudeDelta;
import ch.bailu.aat.gpx.AutoPause;
import ch.bailu.aat.gpx.GpxList;
import ch.bailu.aat.gpx.GpxListWalker;
import ch.bailu.aat.gpx.GpxPoint;
import ch.bailu.aat.gpx.GpxPointFirstNode;
import ch.bailu.aat.gpx.GpxPointNode;
import ch.bailu.aat.gpx.GpxSegmentNode;
import ch.bailu.aat.gpx.MaxSpeed;

public class TimeStampFixer extends GpxListWalker {


    private GpxList newList;
    private boolean newSegment = true;


    private long previousTimeStamp =0;


    @Override
    public boolean doList(GpxList track) {
        newList = new GpxList(track.getDelta().getType(),
                MaxSpeed.NULL,  AutoPause.NULL, AltitudeDelta.NULL);
        return true;
    }


    @Override
    public boolean doSegment(GpxSegmentNode segment) {
        previousTimeStamp = 0;
        newSegment = true;
        return true;
    }

    @Override
    public boolean doMarker(GpxSegmentNode marker) {
        return true;
    }

    @Override
    public void doPoint(GpxPointNode point) {
        if (hasNoErrors(point)) {
            if (newSegment) {
                newSegment = false;
                newList.appendToNewSegment(new GpxPoint(point), point.getAttributes());

            } else {
                newList.appendToCurrentSegment(new GpxPoint(point), point.getAttributes());
            }

            previousTimeStamp = point.getTimeStamp();
        }
    }


    private final static long FIVE_MINUTES = 1000 * 60 * 5;

    private boolean hasNoErrors(GpxPointNode point) {

        if (timeMoreThanPrevious(point)) {
            if (isLastInSegment(point)) {
                return timeNoSkip(point, FIVE_MINUTES);
            } else {
                return timeLessThanNext(point, (GpxPointNode) point.getNext());
            }

        }

        return false;
    }




    private boolean timeMoreThanPrevious(GpxPointNode point) {
        return previousTimeStamp < point.getTimeStamp();
    }


    private boolean timeLessThanNext(GpxPointNode point, GpxPointNode next) {
         return point.getTimeStamp() < next.getTimeStamp();
    }


    private boolean timeNoSkip(GpxPointNode point, long time) {
        return (previousTimeStamp + time) > point.getTimeStamp();
    }

    private boolean isLastInSegment(GpxPointNode point) {
        return point.getNext() == null || point.getNext() instanceof GpxPointFirstNode;
    }

    public GpxList getNewList() {
        return newList;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/tools/TimeStampFixer.kt
package ch.bailu.aat_lib.gpx.tools

import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.gpx.GpxListWalker
import ch.bailu.aat_lib.gpx.GpxPoint
import ch.bailu.aat_lib.gpx.GpxPointFirstNode
import ch.bailu.aat_lib.gpx.GpxPointNode
import ch.bailu.aat_lib.gpx.GpxSegmentNode
import ch.bailu.aat_lib.gpx.attributes.GpxListAttributes

class TimeStampFixer : GpxListWalker() {
    var newList: GpxList? = null
        private set
    private var newSegment = true
    private var previousTimeStamp: Long = 0
    override fun doList(track: GpxList): Boolean {
        newList = GpxList(track.getDelta().getType(), GpxListAttributes.NULL)
        return true
    }

    override fun doSegment(segment: GpxSegmentNode): Boolean {
        previousTimeStamp = 0
        newSegment = true
        return true
    }

    override fun doMarker(marker: GpxSegmentNode): Boolean {
        return true
    }

    override fun doPoint(point: GpxPointNode) {
        if (hasNoErrors(point)) {
            if (newSegment) {
                newSegment = false
                newList?.appendToNewSegment(GpxPoint(point), point.getAttributes())
            } else {
                newList?.appendToCurrentSegment(GpxPoint(point), point.getAttributes())
            }
            previousTimeStamp = point.getTimeStamp()
        }
    }

    private fun hasNoErrors(point: GpxPointNode): Boolean {
        return if (timeMoreThanPrevious(point)) {
            if (isLastInSegment(point)) {
                timeNoSkip(point, FIVE_MINUTES)
            } else {
                timeLessThanNext(point, point.next as GpxPointNode?)
            }
        } else false
    }

    private fun timeMoreThanPrevious(point: GpxPointNode): Boolean {
        return previousTimeStamp < point.getTimeStamp()
    }

    private fun timeLessThanNext(point: GpxPointNode, next: GpxPointNode?): Boolean {
        return point.getTimeStamp() < next!!.getTimeStamp()
    }

    private fun timeNoSkip(point: GpxPointNode, time: Long): Boolean {
        return previousTimeStamp + time > point.getTimeStamp()
    }

    private fun isLastInSegment(point: GpxPointNode): Boolean {
        return point.next == null || point.next is GpxPointFirstNode
    }

    companion object {
        private const val FIVE_MINUTES = (1000 * 60 * 5).toLong()
    }
}
